lib/portlayer/exec/base.go

```diff
@@ -231,7 +233,37 @@ func (c *containerBase) stop(ctx context.Context, waitTime *int32) error {
 
 	log.Warnf("stopping %s via hard power off due to: %s", c.ExecConfig.ID, err)
 
-	return c.poweroff(ctx)
+	err = c.poweroff(ctx)
+
+	// In case of InvalidPowerState and GenericVmConfigFault during hard poweroff, nil would be returned.
+	// At the time the VM must've been already powered off. But we may need to wait for extra time
+	// such that we can obtain the latest powerstate from the VC.
+	if err == nil {
+		timeout, cancel := context.WithTimeout(ctx, defaultPowerStateWaitTime)
+		defer cancel()
+
+		operation := func() error {
+			return c.vm.WaitForPowerState(timeout, types.VirtualMachinePowerStatePoweredOff)
+		}
+
+		isTimeoutErr := func(err error) bool {
+			return timeout.Err() != nil
```

Comment:
You should not retry here if `timeout` was canceled. Only retry if `Err() == DeadlineExceeded`.